src/playwrightTestServer.ts

```diff
@@ -298,14 +307,14 @@ export class PlaywrightTestServer {
     return await testServer.findRelatedTestFiles({ files });
   }
 
-  private _testServer() {
-    if (this._testServerPromise)
-      return this._testServerPromise;
-    this._testServerPromise = this._createTestServer();
-    return this._testServerPromise;
+  private async _testServer() {
+    if (!this._testServerPromise)
+      this._testServerPromise = this._createTestServer();
+    const [server] = await this._testServerPromise;
+    return server;
   }
 
-  private async _createTestServer(): Promise<TestServerConnection | null> {
+  private async _createTestServer(): Promise<[TestServerConnection, string] | []> {
```

Comment:
Don't do return tuples, keep the state in the PlaywrightTestServer instance, you should not need to expose it further.